File: emulinker/src/main/java/org/emulinker/net/PrivateUDPServer.java
package org.emulinker.net;

import com.codahale.metrics.MetricRegistry;
import com.google.common.flogger.FluentLogger;
import java.net.*;
import java.nio.ByteBuffer;
import org.emulinker.util.*;

public abstract class PrivateUDPServer extends UDPServer {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private InetAddress remoteAddress;
  private InetSocketAddress remoteSocketAddress;

  public PrivateUDPServer(
      boolean shutdownOnExit, InetAddress remoteAddress, MetricRegistry metrics) {
    super(shutdownOnExit, metrics);
    this.remoteAddress = remoteAddress;
  }

  public InetAddress getRemoteInetAddress() {
    return remoteAddress;
  }

  public InetSocketAddress getRemoteSocketAddress() {
    return remoteSocketAddress;
  }

  @Override
  protected void handleReceived(ByteBuffer buffer, InetSocketAddress inboundSocketAddress) {
    if (remoteSocketAddress == null) remoteSocketAddress = inboundSocketAddress;
    else if (!inboundSocketAddress.equals(remoteSocketAddress)) {
      logger.atWarning().log(
          "Rejecting packet received from wrong address: "
              + EmuUtil.formatSocketAddress(inboundSocketAddress)
              + " != "
              + EmuUtil.formatSocketAddress(remoteSocketAddress));
      return;
    }

    handleReceived(buffer);
  }

  protected abstract void handleReceived(ByteBuffer buffer);

  protected void send(ByteBuffer buffer) {
    super.send(buffer, remoteSocketAddress);
  }
}

File: emulinker/src/main/java/org/emulinker/net/PrivateUDPServer.kt
package org.emulinker.net

import com.codahale.metrics.Counter
import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.Timer
import com.google.common.flogger.FluentLogger
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import javax.inject.Named
import org.emulinker.util.EmuUtil.formatSocketAddress

abstract class PrivateUDPServer(
  shutdownOnExit: Boolean,
  val remoteInetAddress: InetAddress,
  metrics: MetricRegistry,
  @Named("listeningOnPortsCounter") listeningOnPortsCounter: Counter
) : UDPServer(shutdownOnExit, metrics, listeningOnPortsCounter) {

  private val clientRequestTimer: Timer

  var remoteSocketAddress: InetSocketAddress? = null
    private set

  override fun handleReceived(buffer: ByteBuffer, remoteSocketAddress: InetSocketAddress) {
    if (this.remoteSocketAddress == null) this.remoteSocketAddress = remoteSocketAddress
    else if (remoteSocketAddress != this.remoteSocketAddress) {
      logger
        .atWarning()
        .log(
          "Rejecting packet received from wrong address: %s != %s",
          formatSocketAddress(remoteSocketAddress),
          formatSocketAddress(this.remoteSocketAddress!!)
        )
      return
    }
    clientRequestTimer.time().use { handleReceived(buffer) }
  }

  protected abstract fun handleReceived(buffer: ByteBuffer)

  protected fun send(buffer: ByteBuffer?) {
    super.send(buffer, remoteSocketAddress)
  }

  init {
    clientRequestTimer = metrics.timer(MetricRegistry.name(this.javaClass, "UdpClientRequests"))
  }

  companion object {
    private val logger = FluentLogger.forEnclosingClass()
  }
}
